indico/modules/events/api.py

```diff
@@ -249,274 +494,103 @@ def _get_query_options(self, detail_level):
 
     def category(self, idlist):
         self._detail_level = get_query_parameter(request.args.to_dict(), ['d', 'detail'])
-        filter = None
-        if self._room or self._location or self._eventType:
-            def filter(obj):
-                if self._eventType and obj.getType() != self._eventType:
-                    return False
-                if self._location:
-                    name = obj.getLocation() and obj.getLocation().getName()
-                    if not name or not fnmatch.fnmatch(name.lower(), self._location.lower()):
-                        return False
-                if self._room:
-                    name = obj.getRoom() and obj.getRoom().getName()
-                    if not name or not fnmatch.fnmatch(name.lower(), self._room.lower()):
-                        return False
-                return True
-
         query = (Event.query
                  .filter(~Event.is_deleted,
                          Event.category_chain.overlap(map(int, idlist)),
                          Event.happens_between(self._fromDT, self._toDT))
                  .options(*self._get_query_options(self._detail_level)))
-        return self._process((x.as_legacy for x in query), filter)
+        query = self._update_query(query)
+        return self.serialize_events(x for x in query if x.can_access(self.user) and self._filter_event(x))
```

Comment:
the filter check is probably cheaper than the access check (especially in case of a protected/inheriting event) so i'd switch the two checks here